[chore] Update event semantic conventions (#880)
open-telemetry/semantic-conventions: docs/general/events.md

This document describes the characteristics of standalone Events that are represented
in the data model by `LogRecord`s.

> Semantically, an Event is a named occurrence at an instant in time. It signalsthat

Reviewer: ```suggestion
Semantically, an Event is a named occurrence at an instant in time. It signals that
```

  and the semantic conventions will define the expected structure of the _payload_
  (data) for the event.
* The _payload_ (data) SHOULD be used to represent the structure of the event.

Recommendations for defining events:

* Use the _payload_ (data) to represent the details of the event instead of a
  collection of [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.31.0/specification/common#attribute)
  attributes.
* Events SHOULD be generated / produced / recorded using the
    [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md)
    to ensure that the event is created using the configured SDK instance.
  * The Event API is not yet available in all OpenTelemetry SDKs.
  * TODO: Add deep link to the [compliance matrix of the Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)
    when it exists.
* It's NOT RECOMMENDED to prefix the _payload_ (data) _fields_ with the `event.name` to
  avoid redundancy and to keep the event definition clean.

> * The events SHOULD document their semantic conventions including event name,

Author: SeverityNumber is optional, but I can add.

> * The events SHOULD document their semantic conventions including event name,

Reviewer: could you please create an issue or follow up in a different PR? This one is ready to merge.

bespoke user semantics. When an Event name exists in the semantic conventions, its _payload_
structure and semantics will also be defined.

## Event definition

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| `event.name` | string | Identifies the class / type of event. [1] | `browser.mouse.click`; `device.app.lifecycle` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

**[1]:** Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.31.0/specification/common/attribute-naming.md). Notably, event names are namespaced to avoid collisions and provide a clean separation of semantics for events in separate domains like browser, mobile, and kubernetes.
<!-- endsemconv -->

### General event semantics

* An event MUST have an `event.name` attribute that uniquely identifies the event.

> * It MAY have other [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute)

Reviewer: please set link to a specific tag of semconv. Currently it's `tps://github.com/open-telemetry/opentelemetry-specification/blob/v1.31.0/specification...` and it's updated automatically

  attributes that provide additional context about the event.
* It MAY contain a _payload_ (data) that describes the specific details of the
  named event.
* The structure of the named event (`event.name`) is unique to the event, this
  identifies
  * Whether the event contains a _payload_ (data), and the structure / type of
    the _payload_ (data).

>   * Whether there are any expected [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute)

Reviewer: same here (permalink) and in other places in this doc

    attributes that are used to describe the event.
* All OpenTelemetry defined events SHOULD be defined in the semantic conventions
  to highlight the expected structure of the event.
* The _payload_ (data) MAY contain any type supported by the OpenTelemetry data
  model for the log [body](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body)
  and the semantic conventions will define the expected structure of the _payload_
  (data) for the event.

> * The _payload_ (data) SHOULD be used to represent the details.

Reviewer: the sentence looks incomplete (the details of what?)

> * Events with the same `event.name` are structurally similar to one another and

Reviewer: How it's related to the previous point
> The structure of the named event (`event.name`) is unique to the event, this
  identifies
?

Seems this point ("Events with the same `event.name` are structurally similar to one another") is softer than the previous one

Reviewer 2: Agreed. It seems like we are hanging the definition of the body on the event name so this should be a strong `MUST` statement.

Author: I'm going to remove this item as I was trying to say something like.
Depending on the event each "field" within the payload MAY (or MAY not) be comparable to other events of the same name. But that is getting wordy and really doesn't add anything to the semantic convention as it will depend on the actual event domain on whether it's fields are comparable or not.

> * Events with the same `event.name` are structurally similar to one another and

Reviewer: Would it be possible to define structurally similar? 
What difference is allowed? I assume it's ok to add new field, remove recommended/opt-in field, skip fields with default values (`null`) and any other differences you'd expect a slightly relaxed (de)serializer would allow?

Author: Yeah, I removed this as it's really the domain of the specific event to determine what is and is not allowed. From the basic event structure that is being described here this is an unnecessary clarification as it does not add anything to the convention to say that it is support for an event to evolve and change its structure.

  MAY (depending on the event) be comparable.

Recommendations for defining events:

* Use the _payload_ (data) to represent the details of the event instead of a

>   collection of  over [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute)

Reviewer: ```suggestion
  collection of [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute)
```

  attributes.

> * Generators of events SHOULD use the [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md)

Reviewer: By generators do you mean the instrumentations?

Author: Maybe "producers" would be a better word as it won't just be instrumentations that we create events.

Reviewer: maybe something along the lines of "Events SHOULD be generated/produced/recorded with Event API" would work?

> * Generators of events SHOULD use the [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md)

Reviewer: Would it be possible to add a TODO/comment/note that mentions that event API is not necessarily available in every language yet and add a link to the spec matrix to see which languages support it?

  to create and emit events.
  * While the data model is based on `LogRecord`, it should NOT be assumed that
    ALL consumers will process events as `LogRecords`, some may choose to provide

>     their own Event API implementation.

Reviewer: I don't understand what this sentence means in scope of semantic conventions, would it be possible to remove or clarity it?

* The _payload_ (data) _fields_ SHOULD not be prefixed with the `event.name` to
  avoid redundancy and to keep the event definition clean.

### Event payload (data)

* The yaml definition for OpenTelemetry defined events will support describing
  individual _fields_

>   * Any _fields_ are NOT added to the Global attribute registry for use by other

Reviewer 3: That sentence sounds like a repetition : `are NOT added to the Global attribute registry ... and are NOT added to the Global attribute registry.`

    events and are NOT added to the Global attribute registry.
  * The _fields_ are unique to the named event (`event.name`) and different events
    may use the same _field_ name to represent different data, due to the unique
    nature of the event.
* The _fields_ MAY reference / inherit details from Global attribute registry
  attributes and provide additional details specific to the event, including

>   providing an _alias_ (shorter) name for the attribute.

Reviewer: not blocking: it's not implemented today and I find it confusing having spec in a future tense. Given we already documented that event payload fields are NOT OTel attributes, I'd refer to remove this paragraph or replace it with something like "Common attribute naming conventions and [registry](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/README.md) requirements don't apply to event payload fields."

bespoke user semantics. When an Event name exists in the semantic conventions, its _payload_
structure and semantics will also be defined.

## Event definition

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|

> | [`event.name`](../attributes-registry/event.md) | string | Identifies the class / type of event. [1] | `browser.mouse.click`; `device.app.lifecycle` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Author: I left this for now, as mentioned it's probably not for this PR.

**[1]:** Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.31.0/specification/common/attribute-naming.md). Notably, event names are namespaced to avoid collisions and provide a clean separation of semantics for events in separate domains like browser, mobile, and kubernetes.
<!-- endsemconv -->

### General event semantics

* An event MUST have an `event.name` attribute that uniquely identifies the event.
* It MAY have other [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.31.0/specification/common#attribute)
  attributes that provide additional context about the event.
* It MAY contain a _payload_ (data) that describes the specific details of the
  named event.

> * The structure of the named event (`event.name`) is unique to the event, this
>   identifies
>   * Whether the event contains a _payload_ (data), and the structure / type of
>     the _payload_ (data).
>   * Whether there are any expected [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.31.0/specification/common#attribute)

Reviewer: ```suggestion
* The event name uniquely identifies event structure / type of the _payload_ (data) and the set of attributes.
```

    attributes that are used to describe the event.

> * All OpenTelemetry defined events SHOULD be defined in the semantic conventions

Author: While instrumentations are probably the main source of events, they may not be the only generator, I'll reword to incorporate the details.

And yes, moving to the recommendations section makes sense

  to highlight the expected structure of the event.
* The _payload_ (data) MAY contain any type supported by the OpenTelemetry data
  model for the log [body](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body)
  and the semantic conventions will define the expected structure of the _payload_
  (data) for the event.

> * The _payload_ (data) SHOULD be used to represent the structure of the event.

Reviewer: Do we still need this sentence? What new information does it convey that's not covered before?

Author: We are trying to convey that when defining events it is recommended that you should use the payload to describe the detail of the event vs using a collection of standard attributes.

It's a should because there are some existing legacy cases (exceptions are a good example) where it would probably be too much of a breaking change (or too much additional confusion) to introduce yet another way to represent exceptions that are not directly compatible.

Recommendations for defining events:

* Use the _payload_ (data) to represent the details of the event instead of a
  collection of [standard](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.31.0/specification/common#attribute)
  attributes.
* Events SHOULD be generated / produced / recorded using the
    [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md)
    to ensure that the event is created using the configured SDK instance.
  * The Event API is not yet available in all OpenTelemetry SDKs.
  * TODO: Add deep link to the [compliance matrix of the Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)
    when it exists.

> * The _payload_ (data) _fields_ SHOULD not be prefixed with the `event.name` to

Reviewer: maybe ?
```suggestion
* It's NOT RECOMMENDED to prefix the _payload_ (data) _fields_ with the `event.name` to
```

Otherwise, please update `SHOULD not` to `SHOULD NOT`

  avoid redundancy and to keep the event definition clean.

### Event payload (data)

* Common attribute naming conventions and [registry](../attributes-registry/README.md)
  requirements don't apply to event payload fields.
* The definition for OpenTelemetry defined events supports describing
  individual _fields_ (Body Fields)

>   * Any _fields_ are NOT added to the Global attribute registry for use by other

Reviewer: do we need this sentence? 

it's covered in the "Common attribute naming conventions and [registry](../attributes-registry/README.md)"

    events.
  * The _fields_ are unique to the named event (`event.name`) and different events
    may use the same _field_ name to represent different data, due to the unique
    nature of the event.

> * The _fields_ MAY reference / inherit details from Global attribute registry

Reviewer: ```suggestion
* The _fields_ MAY reference / inherit details from the attribute registry
```

  attributes and provide additional details specific to the event, including
  providing an _alias_ (shorter) name for the attribute.

As each `event.name` is unique and defines the structure of the event, as long
as the event follows the General event semantics it will be considered a valid

> event. For example :

Reviewer: > As each `event.name` is unique and defines the structure of the event, as long
as the event follows the General event semantics it will be considered a valid
event.

Does the exception logs referenced below demonstrate something related to this sentence? 

Also, what information do we share in this sentence? A definition of the valid event? But we already defined it above.
Unless there is a notion of valid or invalid event or some handling that needs to happen for invalid ones, I suggest removing this sentence.

Author: Yeah, I think we can clarify this later if needed, as it not really adding much as it is today.

In terms of "invalid" events, its really going to be up to each event definition on what / how it should be handled (if) an event introduces breaking changes within it's own domain.
